cmd/juju/config/config.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package config
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/cmd/v3"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/utils/keyvalues"
+)
+
+// ConfigAction represents the action we want to perform here.
+// The action will be set in Init, and then accessed by the child command's
+// Run method to decide what to do.
+type ConfigAction string
+
+const (
+	GetOne  ConfigAction = "get"
+	GetAll  ConfigAction = "get all"
+	Set     ConfigAction = "set"
+	SetFile ConfigAction = "set from file"
+	Reset   ConfigAction = "reset"
+)
+
+// ConfigCommandBase provides a common interface/functionality for configuration
+// commands (such as config and model-config). It defines SetFlags and Init
+// methods, while the individual command needs to define its own Run and Info
+// methods.
+type ConfigCommandBase struct {
+	Out cmd.Output // TODO: put this in child command?
+
+	// Fields to be set by child
+	CantReset []string // keys which can't be reset
+
+	// Flag values
+	ConfigFile cmd.FileVar
+	reset      []string // Holds the keys to be reset until parsed.
+
+	// Fields to be set during initialisation
+	Actions     []ConfigAction // The action which we want to handle, set in Init.
+	KeyToGet    string
+	KeysToReset []string // Holds keys to be reset after parsing.
+	ValsToSet   map[string]string
+}
+
+// Info - to be implemented by child command
+
+// SetFlags implements cmd.Command.SetFlags.
+func (c *ConfigCommandBase) SetFlags(f *gnuflag.FlagSet) {
+	// Set the --format and -o flags
+	c.Out.AddFlags(f, "yaml", map[string]cmd.Formatter{
+		"json": cmd.FormatJson,
+		"yaml": cmd.FormatYaml,
+		// "tabular":     formatConfigTabular,
```

Comment:
Commented-out code. (Though I suspect this is going to go anyway seeing you're moving `Out`.)

In general I think it's best to remove these before review, otherwise it's hard for the reviewer to tell whether you intended to delete them or uncomment them. (And if the reviewer wants to see the original, it's there in the diff.)